File: data-engineering/python/spark210/apps/wordcount.py
#!/usr/bin/env python3
# -*- coding: utf8 -*-


import sys
from operator import add
from pyspark import SparkConf, SparkContext


class WordCount:
    def get_spark_context(self, app_name, master):
        conf = SparkConf().setAppName(app_name).setMaster(master)

        return SparkContext(conf=conf)

    def get_input_rdd(self, sc, input):
        return sc.textFile(input)

    def process(self, input_rdd):
        words = input_rdd.flatMap(lambda s : s.split(" "))
        wc_pair = words.map(lambda s: (s, 1))
        return wc_pair.reduceByKey(add)


if __name__ == "__main__":
    wc = WordCount()
    sc = wc.get_spark_context("WordCount", sys.argv[1])

    sc.setLogLevel("ERROR")

    input_rdd = wc.get_input_rdd(sc, sys.argv[2])
    result_rdd = wc.process(input_rdd)

    result_rdd.saveAsTextFile(sys.argv[3])
    sc.stop()

File: data-engineering/scala/spark210/src/main/scala/apps/WordCount.scala
package apps

import org.apache.spark.rdd.RDD
import org.apache.spark.{SparkConf, SparkContext}

object WordCount {
  def getSparkContext(appName: String, master: String) = {
    val conf = new SparkConf().setAppName(appName).setMaster(master)
    new SparkContext(conf)
  }

  def getInputRDD(sc: SparkContext, input: String) = {
    sc.textFile(input)
  }

  def process(inputRDD: RDD[String]) = {
    val words = inputRDD.flatMap(str => str.split(" "))
    val wcPair = words.map((_, 1))

    wcPair.reduceByKey(_ + _)
  }

  def handleResult(resultRDD: RDD[(String, Int)], output: String): Unit = {
    resultRDD.saveAsTextFile(output);
  }

  def main(args: Array[String]): Unit = {
    require(args.length == 3, "Usage: WordCount <Master> <Input> <Output>")

    val sc = getSparkContext("WordCount", args(0))
    val inputRDD = getInputRDD(sc, args(1))
    val resultRDD = process(inputRDD)

    handleResult(resultRDD, args(2))
  }
}
